File: 백준/Gold/3190. 뱀/뱀.py
import sys
input = sys.stdin.readline

n = int(input())
k = int(input()) # 사과 개수
board = [[0]* (n+1) for _ in range(n+1)]

for _ in range(k):
    x, y = map(int, input().split())
    board[x][y] = 2

l = int(input()) # 방향 변환 횟수
time = []
direction = []

for _ in range(l):
    x, c = input().split()
    time.append(x)
    direction.append(c)

x, y = 1, 1
board[x][y] = 1
dx, dy = 0, 1
next_time = int(time.pop(0))
next_d = direction.pop(0)
cur_time = 0
snake = [(x,y)]

while True:
    # 시간되면 방향 바꾸기
    if cur_time == next_time:
        if next_d == 'L':
            dx, dy = -1 * dy, dx
        else:
            dx, dy = dy, -1 * dx
        if len(time) != 0:
            next_time = int(time.pop(0))
            next_d = direction.pop(0)

    # 한 칸씩 이동하기
    nx, ny = x + dx, y + dy

    # 맵 밖에 나가면 탈출
    if nx <= 0 or nx > n or ny <= 0 or ny > n:
        break
    # 뱀에 닿으면 탈출
    if board[nx][ny] == 1:
        break

    if board[nx][ny] != 2: # 사과 없으면 꼬리 이동
        px, py = snake.pop(0)
        board[px][py] = 0
        
    # 머리 이동
    board[nx][ny] = 1
    snake.append((nx, ny))
    
    # 1초 증가
    cur_time += 1
    x, y = nx, ny

print(cur_time + 1)

File: 백준/Gold/3190. 뱀/뱀.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n");

let index = 0;
const n = Number(input[index++]);
const board = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));

const k = Number(input[index++]);
for (let i = 0; i < k; i++) {
  const [x, y] = input[index++].split(" ").map(Number);
  board[x][y] = "A";
}

const l = Number(input[index++]);
const turns = {};
for (let i = 0; i < l; i++) {
  const [time, direction] = input[index++].split(" ");
  turns[time] = direction;
}

const direction = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
]; // 우하좌상
let d = 0; // 우측 방향으로 시작

console.log(move());

function move() {
  let time = 1;
  let x = 1;
  let y = 1;
  const snake = [{ x, y }];
  board[x][y] = "S";

  while (true) {
    const nx = x + direction[d][0];
    const ny = y + direction[d][1];

    if (nx <= 0 || ny <= 0 || nx > n || ny > n) return time;

    if (board[nx][ny] === "S") {
      return time;
    }

    if (board[nx][ny] !== "A") {
      const tail = snake.shift();
      board[tail.x][tail.y] = 0;
    }

    board[nx][ny] = "S";
    snake.push({ x: nx, y: ny });

    for (let [t, dir] of Object.entries(turns)) {
      if (time === Number(t)) {
        if (dir === "L") d = d - 1;
        else d = d + 1;

        if (d < 0) d += 4;
        if (d >= 4) d -= 4;
      }
    }

    time++;
    x = nx;
    y = ny;
  }
}
